File: worker/utils/qstash_verify.py
import hashlib
import hmac
import os
from typing import Optional

from fastapi import Request


async def verify_qstash_signature(request: Request) -> bool:
    """Verify QStash webhook signature."""
    signature = request.headers.get("upstash-signature")
    if not signature:
        return False

    current_key = os.environ.get("QSTASH_CURRENT_SIGNING_KEY")
    next_key = os.environ.get("QSTASH_NEXT_SIGNING_KEY")

    if not current_key or not next_key:
        return False

    body = await request.body()

    for key in [current_key, next_key]:
        expected = hmac.new(
            key.encode(), body, hashlib.sha256
        ).hexdigest()
        if hmac.compare_digest(signature, expected):
            return True

    return False

File: supabase/functions/_shared/qstash-verify.ts
import { Receiver } from "https://esm.sh/@upstash/qstash@2.7.17";

let receiver: Receiver | null = null;

function getReceiver(): Receiver {
  if (!receiver) {
    const currentKey = Deno.env.get("QSTASH_CURRENT_SIGNING_KEY");
    const nextKey = Deno.env.get("QSTASH_NEXT_SIGNING_KEY");
    if (!currentKey || !nextKey) {
      throw new Error("Missing QStash signing keys");
    }
    receiver = new Receiver({ currentSigningKey: currentKey, nextSigningKey: nextKey });
  }
  return receiver;
}

export async function verifyQStashSignature(req: Request): Promise<boolean> {
  const signature = req.headers.get("upstash-signature");
  if (!signature) return false;

  try {
    const body = await req.clone().text();
    await getReceiver().verify({ signature, body });
    return true;
  } catch {
    return false;
  }
}
